lib/stripe/gateway.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require 'stripe'
+require 'stripe/input_data_validator'
+require 'stripe/result'
+
+module Stripe
+  class Gateway
+    def self.charge(data)
+      new(data).charge
+    end
+
+    def initialize(data)
+      @data = data
+      @api_errors = []
+    end
+
+    def charge
+      Result.new(charge_result, errors)
+    end
+
+    private
+
+    attr_reader :data, :api_errors
+
+    def charge_result
+      charge_and_rescue_data_related_errors
+    rescue Stripe::RateLimitError
+      save_error(:too_many_requests)
+    rescue Stripe::APIConnectionError
+      save_error(:connection_problems)
+    rescue Stripe::StripeError
+      save_error(:stripe_error)
+    rescue StandardError
+      save_error(:unknown_error)
```

Comment:
I would do something with your errors that also allows you to store the exception. This will be very hard to debug in the future if all you have is `:unknown_error`

Reply:
Changing your data object to be an array of `Error = Struct.new(:code, :data)` which probably suffice?

Reply:
I think this will happen in a separate branch when we're forced to think about logging and what information we will need.

Reply:
yeah, at the moment absolutely everything is returning just an array of symbols. But it is written in a way that it will be easy to change in the future, which is what matters.